File: 3756-concatenate-non-zero-digits-and-multiply-by-sum-ii/3756-concatenate-non-zero-digits-and-multiply-by-sum-ii.cpp
class Solution {
public:
    static const int MOD = 1e9+7;
    vector<int> sumAndMultiply(string s, vector<vector<int>>& queries) {
        int n = (int)s.size();

        vector<int> nonZeroCount(n, 0);
        vector<int> numberUpTo(n, 0);
        vector<int> digitSumUpTo(n, 0);
        vector<int> pow10(n + 1, 0);

        pow10[0] = 1;
        for (int i = 1; i <= n; ++i) {
            pow10[i] = (int)((1LL * pow10[i - 1] * 10) % MOD);
        }

        nonZeroCount[0] = (s[0] != '0') ? 1 : 0;
        numberUpTo[0] = s[0] - '0';
        digitSumUpTo[0] = s[0] - '0';

        for (int i = 1; i < n; ++i) {
            int digit = s[i] - '0';
            nonZeroCount[i] = nonZeroCount[i - 1] + (digit != 0 ? 1 : 0);
        }

        for (int i = 1; i < n; ++i) {
            int digit = s[i] - '0';
            if (digit != 0) {
                numberUpTo[i] = (int)((1LL * numberUpTo[i - 1] * 10 + digit) % MOD);
            } else {
                numberUpTo[i] = numberUpTo[i - 1];
            }
        }

        for (int i = 1; i < n; ++i) {
            digitSumUpTo[i] = digitSumUpTo[i - 1] + (s[i] - '0');
        }

        int q = (int)queries.size();
        vector<int> result(q, 0);

        for (int i = 0; i < q; ++i) {
            int l = queries[i][0];
            int r = queries[i][1];

            int startCount = (l == 0) ? 0 : nonZeroCount[l - 1];
            int numBefore = (l == 0) ? 0 : numberUpTo[l - 1];

            int endCount = nonZeroCount[r];
            int subStrLen = endCount - startCount;

            if (subStrLen == 0) {
                result[i] = 0;
                continue;
            }

            int x = (numberUpTo[r] - (int)((1LL * numBefore * pow10[subStrLen]) % MOD) + MOD) % MOD;
            int sum_digits = digitSumUpTo[r] - ((l == 0) ? 0 : digitSumUpTo[l - 1]);

            result[i] = (int)((1LL * x * sum_digits) % MOD);
        }

        return result;
    }
};

File: 3756-concatenate-non-zero-digits-and-multiply-by-sum-ii/3756-concatenate-non-zero-digits-and-multiply-by-sum-ii.py
class Solution:
    MOD = int(1e9 + 7)
    def sumAndMultiply(self, s: str, queries: List[List[int]]) -> List[int]:
        n = len(s)

        nonZeroCount = [0] * n
        numberUpTo = [0] * n
        digitSumUpTo = [0] * n
        pow10 = [0] * (n + 1)

        pow10[0] = 1
        for i in range(1, n + 1):
            pow10[i] = (pow10[i - 1] * 10) % self.MOD

        nonZeroCount[0] = 1 if s[0] != '0' else 0
        numberUpTo[0] = int(s[0])
        digitSumUpTo[0] = int(s[0])

        for i in range(1, n):
            digit = int(s[i])
            nonZeroCount[i] = nonZeroCount[i - 1] + (1 if digit != 0 else 0)

        for i in range(1, n):
            digit = int(s[i])
            if digit != 0:
                numberUpTo[i] = (numberUpTo[i - 1] * 10 + digit) % self.MOD
            else:
                numberUpTo[i] = numberUpTo[i - 1]

        for i in range(1, n):
            digitSumUpTo[i] = digitSumUpTo[i - 1] + int(s[i])

        q = len(queries)
        result = [0] * q

        for i in range(q):
            l, r = queries[i]

            startCount = 0 if l == 0 else nonZeroCount[l - 1]
            numBefore = 0 if l == 0 else numberUpTo[l - 1]

            endCount = nonZeroCount[r]
            subStrLen = endCount - startCount

            if subStrLen == 0:
                result[i] = 0
                continue

            x = (numberUpTo[r] - (numBefore * pow10[subStrLen] % self.MOD) + self.MOD) % self.MOD
            sum_digits = digitSumUpTo[r] - (0 if l == 0 else digitSumUpTo[l - 1])

            result[i] = (x * sum_digits) % self.MOD

        return result